### PR TITLE
Datasources: provide generic function to extract custom headers

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
**What is this feature?**

This feature provided a generic function to extract custom headers from datasources.

**Why do we need this feature?**

In the alerting team we remarked that there was no way of doing this. We think it adds value for others that want to do the same.

**Who is this feature for?**

This PR is a subset of #63819

### CODE REVIEW

**Comment by JohnnyQQQQ** [comment]
If someone from @grafana/plugins-platform-backend could review it, it would be highly appreciated.

**Review by gotjosh — APPROVED** [review]
LGTM - godspeed mate.

**Comment by marefr** [comment]
Sure we'll have a look thanks. How are you planning to use this from within alerting, can you give an example?

**Comment by JohnnyQQQQ** [comment]
It's basically extracting this as a genric function https://github.com/grafana/grafana/pull/63819/files#diff-84ce8e0f3f9dd06f7632d969cf6d971cda82dd4e8c22f12c6cbffca9a17ba47eR261-R294

File: pkg/services/datasources/models.go
Comment (marefr) [inline pkg/services/datasources/models.go:107]: First, this can be costly/create performance issues if used in a hot path. Second, I would probably suggest putting this method in the [DataSourceService](https://github.com/grafana/grafana/blob/main/pkg/services/datasources/datasources.go#L14) since there's already a `getCustomHeaders` method in there.

**Review by marefr — COMMENTED** [review]
Looks good in general but I suggest a change

**Comment by JohnnyQQQQ** [comment]
@marefr thanks for taking the time to review. I addressed your suggestion.

**Review by gotjosh — APPROVED** [review]
LGTM, but please wait for @marefr approval.

**Comment by JohnnyQQQQ** [comment]
I double-checked the use of the context. It's only used for the [store](https://github.com/grafana/grafana/blob/e7a7f8320f2c9e79d131575f6833fb5fe84c8321/pkg/services/secrets/manager/manager.go#L362) currently and ignored during [decryption](https://github.com/grafana/grafana/blob/e7a7f8320f2c9e79d131575f6833fb5fe84c8321/pkg/services/encryption/provider/decipher_aes.go#L16).

File: pkg/services/datasources/service/datasource.go
Comment (JohnnyQQQQ) [inline pkg/services/datasources/service/datasource.go:655]:
```suggestion
// CustomerHeaders returns the custom headers specified in the datasource. The context is used for the decryption operation that might use the store, so consider setting an acceptable timeout for your use case. 
```

**Review by marefr — APPROVED** [review]
LGTM great

**Pull request merged**
